### PR TITLE
Preserve namespace on imagestreams server-side export

### PR DESCRIPTION
**Pull request by juanvallejo**
Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1542238

cc @deads2k

### CODE REVIEW

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go [hunk 1]
```diff
@@ -1199,9 +1199,6 @@ func (e *Store) calculateTTL(obj runtime.Object, defaultTTL int64, update bool)
 // present when the object is exported.
 func exportObjectMeta(accessor metav1.Object, exact bool) {
 	accessor.SetUID("")
-	if !exact {
-		accessor.SetNamespace("")
```
Comment (juanvallejo) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go]: @deads2k from what I can tell, it wasn't the export strategy that was defaulting an empty namespace to "default".
In the case of the imagestream I tested with, its `ExportStrategy` was nil. Instead, the `CreateStrategy` was [called here](https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L1231). I suspect it is somewhere there that the namespace is defaulted

**Comment by mfojtik** [comment]
/hold

Please have upstream PR approved first and use the upstream PR number in commit message.

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go (hunk 1 above)
Comment (deads2k) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go]: This code looks correct.  Check the return value from teh apiserver.  Was it actually "default" or was that set client-side?

**Comment by deads2k** [comment]
/hold

fix doesn't look correct.

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go (hunk 1 above)
Comment (juanvallejo) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go]: Taking a second look, this problem does not exist when testing against a cluster created with `openshift start`.

Can only replicate if using a cluster created via `oc cluster up`. Still looking

Comment (juanvallejo) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go]: @deads2k 
>  Was it actually "default" or was that set client-side?

It does not look like the client is setting it to "default", from what I can tell.
--loglevel=8 of `oc get is/jenkins -o yaml --export`: http://pastebin.test.redhat.com/553666

Comment (juanvallejo) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go]: @mfojtik @deads2k for comparison, here is `--loglevel 8` output for the same command, but executed against a cluster created via `openshift start` instead: http://pastebin.test.redhat.com/553669

**Comment by mffiedler** [comment]
My current namespace was "logging" not "default" in latest repro in https://bugzilla.redhat.com/show_bug.cgi?id=1542238

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go (hunk 1 above)
Comment (juanvallejo) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go]: cc @bparees I'm not sure if `oc cluster up` uses a different export strategy / why it appears to behave differently than an `$ openshift start` cluster.

We handle server-side export here: https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L1203 clearing the namespace field for an object here https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L1203

Not sure how / where an empty namespace field could be defaulting to `default` as its value. The client does not appear to be altering this, as loglevel output linked to above shows that the cluster is already returning the object with the namespace field value defaulted.

This does not happen with `openshift start` clusters. I have tested with 3.8 and 3.9 releases of the openshift binary.

I have also tested with `oc cluster up --latest` and `oc cluster up --version v3.8.0-alpha.0`. Was hoping you had maybe seen something like this before?

Comment (bparees) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go]: > Was hoping you had maybe seen something like this before?

i haven't, and no, oc cluster up certainly doesn't configure export strategies, so i can't imagine why there would be a difference in behavior.

also "oc export" seems to do the right thing (leaves off the namespace).

Anyway I'm seeing the "wrong" behavior using openshift start clusters, as well as oc cluster up clusters:

exporting from current project==openshift:

```
$ oc get --export is jenkins -o yaml
apiVersion: image.openshift.io/v1
kind: ImageStream
metadata:
  annotations:
    openshift.io/display-name: Jenkins
    openshift.io/image.dockerRepositoryCheck: 2018-02-07T19:19:34Z
  creationTimestamp: null
  generation: 1
  name: jenkins
  namespace: default
```

**Comment by juanvallejo** [comment]
@mfojtik @soltysh @deads2k Updated this PR to just add an export strategy for imagestreams. Before, they did not have one, so doing a server-side export was defaulting to a call to the imagestream create-strategy. This strategy was [defaulting the namespace to the "Default" value](https://github.com/openshift/origin/blob/master/pkg/image/registry/imagestream/strategy.go#L134) when it saw that the namespace field had been cleared. ptal

**Comment by bparees** [comment]
lgtm for whatever that's worth.

**Comment by soltysh** [comment]
I've talked with @juanvallejo on IRC he'll try to fix setting the namespace in `dockerImageRepository` so that it is not set when being called from `PrepareForCreate`.

**Comment by juanvallejo** [comment]
@soltysh thanks for the feedback. Added an extra arg to `dockerImageRepository` to allow/prevent namespace defaulting. PTAL

**Comment by soltysh** [comment]
@juanvallejo update tests:
```
# github.com/openshift/origin/pkg/image/registry/imagestream
[WARNING] pkg/image/registry/imagestream/strategy_test.go:184:42: not enough arguments in call to strategy.dockerImageRepository
[WARNING] 	have (*image.ImageStream)
[WARNING] 	want (*image.ImageStream, bool) 
```
and you're good to go.

**Pull request merged**